internal/data/common.go

```diff
@@ -140,6 +142,136 @@ func NewAttributes(m AttributesMap, droppedCount uint32) Attributes {
 	return Attributes{m, droppedCount}
 }
 
+// StringKeyValue stores a key and AttributeValue pair.
+type StringKeyValue struct {
+	orig *otlpcommon.StringKeyValue
+}
+
+// NewStringKeyValue creates a new StringKeyValue with the given key and value.
+func NewStringKeyValue(k string, v string) StringKeyValue {
+	return StringKeyValue{&otlpcommon.StringKeyValue{Key: k, Value: v}}
+}
+
+// Key returns the key associated with this StringKeyValue.
+func (akv StringKeyValue) Key() string {
+	return akv.orig.Key
+}
+
+// Value returns the value associated with this StringKeyValue.
+func (akv StringKeyValue) Value() string {
+	return akv.orig.Value
+}
+
+// SetValue replaces the value associated with this StringKeyValue.
+func (akv StringKeyValue) SetValue(v string) {
+	akv.orig.Value = v
+}
+
+// StringMap stores a map of attribute keys to values.
+type StringMap struct {
+	orig *[]*otlpcommon.StringKeyValue
+}
+
+func newStringMap(orig *[]*otlpcommon.StringKeyValue) StringMap {
+	return StringMap{orig}
+}
+
+// NewStringMap creates a new StringMap from the given map[string]AttributeValue.
+func NewStringMap(attrMap map[string]string) StringMap {
+	if len(attrMap) == 0 {
+		var orig []*otlpcommon.StringKeyValue
+		orig = nil
+		return StringMap{&orig}
+	}
+	origs := make([]otlpcommon.StringKeyValue, len(attrMap))
+	wrappers := make([]*otlpcommon.StringKeyValue, len(attrMap))
+
+	ix := 0
+	for k, v := range attrMap {
+		wrappers[ix] = &origs[ix]
+		wrappers[ix].Key = k
+		StringKeyValue{wrappers[ix]}.SetValue(v)
+		ix++
+	}
+
+	return StringMap{&wrappers}
+}
+
+// Get returns the StringKeyValue associated with the key and true,
+// otherwise an invalid instance of the StringKeyValue and false.
+func (sm StringMap) Get(k string) (StringKeyValue, bool) {
+	for _, a := range *sm.orig {
+		if a.Key == k {
+			return StringKeyValue{a}, true
+		}
+	}
+	return StringKeyValue{nil}, false
+}
+
+// Delete deletes the entry associated with the key and returns true if the key
+// was present in the map, otherwise returns false.
+func (sm StringMap) Delete(k string) bool {
+	for i, a := range *sm.orig {
+		if a.Key == k {
+			(*sm.orig)[i] = (*sm.orig)[len(*sm.orig)-1]
+			*sm.orig = (*sm.orig)[:len(*sm.orig)-1]
+			return true
+		}
+	}
+	return false
+}
+
+// Insert adds the StringKeyValue to the map when the key does not exist.
+// No action is applied to the map where the key already exists.
+func (sm StringMap) Insert(k, v string) {
+	if _, existing := sm.Get(k); !existing {
+		*sm.orig = append(*sm.orig, NewStringKeyValue(k, v).orig)
```

Comment:
This will use existing capacity in slice of pointers but will always allocate a new instance of `otlpcommon.StringKeyValue` even if there may be room in the slice of `orig` structs. I think we can extend the original `orig` slice but that request unsafe pointer conversion. Probably not worth it.

Reply:
We can always optimize later.